sync/validation_queue.go

```diff
@@ -44,149 +67,146 @@ func (vq *validationQueue) inQueue(id types.BlockID) bool {
+		//continue
+	}
+	if res == false {
+		vq.Info("pending done for %v", blk.ID())
+		vq.updateDependencies(blk.ID(), true)
+	}
+}
 
-		if len(vq.reverseDepMap) == 0 {
-			vq.done()
+func (vq *blockQueue) finishBlockCallback(block *types.Block) func(res bool) error {
+	return func(res bool) error {
+		if !res {
+			vq.Info("finished block %v block invalid", block.ID())
 			return nil
 		}
-	}
-
-	return nil
-}
 
-func (vq *validationQueue) finishBlockCallback(s *Syncer, block *types.Block) func() error {
-	return func() error {
 		//data availability
-		txs, txErr, atxs, atxErr := s.DataAvailability(block)
-		if txErr != nil || atxErr != nil {
-			return fmt.Errorf("txerr %v, atxerr %v", txErr, atxErr)
+		txs, atxs, err := vq.DataAvailabilty(block)
+		if err != nil {
+			return fmt.Errorf("DataAvailabilty failed for block %v err: %v", block, err)
 		}
 
 		//validate block's votes
-		if valid := s.validateVotes(block); valid == false {
+		if valid := validateVotes(block, vq.ForBlockInView, vq.Hdist); valid == false {
 			return errors.New(fmt.Sprintf("validate votes failed for block %v", block.ID()))
 		}
 
-		if err := s.AddBlockWithTxs(block, txs, atxs); err != nil {
+		if err := vq.AddBlockWithTxs(block, txs, atxs); err != nil && err != mesh.ErrAlreadyExist {
 			return err
 		}
 
 		return nil
 	}
 }
 
-func (vq *validationQueue) updateDependencies(block types.BlockID) []types.BlockID {
-	delete(vq.depMap, block)
-	var blocks []types.BlockID
+func (vq *blockQueue) updateDependencies(block types.BlockID, valid bool) {
+	vq.Lock()
+	defer vq.Unlock()
+	var doneBlocks []types.BlockID
 	doneQueue := make([]types.BlockID, 0, len(vq.depMap))
-	doneQueue = vq.removefromDepMaps(block, doneQueue)
+	doneQueue = vq.removefromDepMaps(block, valid, doneQueue)
 	for {
 		if len(doneQueue) == 0 {
-			return blocks
+			break
 		}
 		block = doneQueue[0]
 		doneQueue = doneQueue[1:]
-		blocks = append(blocks, block)
-		doneQueue = vq.removefromDepMaps(block, doneQueue)
+		doneBlocks = append(doneBlocks, block)
+		doneQueue = vq.removefromDepMaps(block, valid, doneQueue)
 	}
-	return nil
 }
 
-func (vq *validationQueue) removefromDepMaps(block types.BlockID, queue []types.BlockID) []types.BlockID {
-	for _, b := range vq.reverseDepMap[block] {
-		delete(vq.depMap[b], block)
-		if len(vq.depMap[b]) == 0 {
-			delete(vq.depMap, b)
-			queue = append(queue, b)
+func (vq *blockQueue) removefromDepMaps(block types.BlockID, valid bool, doneBlocks []types.BlockID) []types.BlockID {
+	//clean after block
+	delete(vq.depMap, block)
+	delete(vq.callbacks, block)
+	delete(vq.visited, block)
+
+	for _, dep := range vq.reverseDepMap[block] {
+		delete(vq.depMap[dep], block)
+		if len(vq.depMap[dep]) == 0 {
+			delete(vq.depMap, dep)
+			vq.Info("run callback for %v, %v", dep, reflect.TypeOf(dep))
+			if callback, ok := vq.callbacks[dep]; ok {
+				if err := callback(valid); err != nil {
+					vq.Error(" %v callback Failed", dep)
```

Comment:
Don't you need to remove the callback from `callbacks` in such case (like you do in the case of success)